### ISSUE TITLE
Compile time shaders translation

### ISSUE THREAD

**Issue by enfipy** [bevyengine/bevy#648]
I noticed that in some scenarios there's no better solution than translate shaders (GLSL or any other) to SPIR-V at compile time (or AOT compilation). For example there's official [doc](https://developer.android.com/ndk/guides/graphics/shader-compilers) about vulkan shaders compilation for Android and that notices about Ahead of time compilation that places final spv files directly to apk.

Also, there is some examples of vulkan + build time compilation of shaders [here](https://www.github.com/adrien-ben/vulkan-tutorial-rs/tree/master/build.rs).

So I'm curious, is it right and could be added to Bevy?

**Comment by cart** [comment]
Yeah Bevy will eventually support pre-compiled shaders. Theres a bit of nuance because Bevy allows using run-time state to dynamically compile specialized shaders, but we could work around that. Once our WIP asset system rework lands, we could make shaders (optionally) precompiled to spirv when the source file is imported.

**Comment by enfipy** [comment]
Sounds good!

Do you have any design thoughts on how pre-compiled shaders should be specified? Most easiest way to add this through `build.rs` but I think it would be much better to somehow make interface almost same as runtime compiled shader assets (maybe with macros or const functions).

**Comment by cart** [comment]
I think it would use the upcoming "asset import" features. You would just drop your shader source into the `assets` folder, then somehow kick off the import process (the bevy editor will do this when it comes out, but until then its something users will need to run manually, maybe at app startup). This process would load the shader source files, compile them, then write the results to the ".import" folder. Games would then load the compiled shader assets from the ".import" folder.

The same process will be used to optimize certain types of assets like Meshes for released games.
